src/libappimage/desktop_integration/Thumbnailer.cpp

```diff
@@ -0,0 +1,156 @@
+
+            try {
+                std::vector<uint8_t> iconData{normalIconData.begin(), normalIconData.end()};
+                IconHandle iconHandle(iconData);
+                iconHandle.setSize(128);
+                iconHandle.save(normalThumbnailPath.string(), "png");
+                return;
+            } catch (const IconHandleError&) {}
+
+            // It wasn't possible to generate a thumbnail, therefore the the icon will be written as it's
+            bf::ofstream out(normalThumbnailPath);
+            out.write(normalIconData.data(), normalIconData.size());
+        }
+
+        void Thumbnailer::generateLargeSizeThumbnail(const std::string& canonicalPathMd5,
+                                                     std::vector<char>& largeIconData) const {
+            bf::path largeThumbnailPath = getLargeThumbnailPath(canonicalPathMd5);
+            bf::create_directories(largeThumbnailPath.parent_path());
+
+            try {
+                std::vector<uint8_t> iconData{largeIconData.begin(), largeIconData.end()};
+                IconHandle iconHandle(iconData);
+                iconHandle.setSize(256);
+                iconHandle.save(largeThumbnailPath.string(), "png");
+                return;
+            } catch (const IconHandleError&) {}
+
+            // It wasn't possible to generate a thumbnail, therefore the the icon will be written as it's
+            bf::ofstream out(largeThumbnailPath);
+            out.write(largeIconData.data(), largeIconData.size());
+            out.close();
+        }
+
+
+        std::string Thumbnailer::getCanonicalPathMd5(const std::string& appImagePath) const {
+            auto canonicalAppImagePath = bf::weakly_canonical(appImagePath).string();
+            auto md5Raw = appimage_get_md5(canonicalAppImagePath.c_str());
+            std::string canonicalPathMd5 = md5Raw ?: "";
+
+            free(md5Raw);
+            return canonicalPathMd5;
+        }
+
+        bf::path Thumbnailer::getNormalThumbnailPath(const std::string& canonicalPathMd5) const {
+            bf::path xdgCacheHomePath(xdgCacheHome);
+
+            bf::path normalThumbnailPath =
+                xdgCacheHomePath / "thumbnails/normal" / (canonicalPathMd5 + ".png");
+            return normalThumbnailPath;
+        }
+
+        bf::path Thumbnailer::getLargeThumbnailPath(const std::string& canonicalPathMd5) const {
+            bf::path xdgCacheHomePath(xdgCacheHome);
+
+            bf::path largetThumbnailPath =
+                xdgCacheHomePath / "thumbnails/large" / (canonicalPathMd5 + ".png");
+            return largetThumbnailPath;
+        }
+
+        std::vector<char> Thumbnailer::getIconData(const DesktopIntegrationResources& resources,
+                                                   const std::string& appIcon, const std::string& iconSize) {
+            std::vector<char> iconData;
+
+            for (const auto& itr: resources.icons) {
```

Comment:
Please add some CSI comment here (not entirely sure what these `if`s are good for).

Reply:
Done